server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocatorTests.java

```diff
@@ -231,117 +213,149 @@ public void afterPrimariesBeforeReplicas(RoutingAllocation allocation) {}
-        }
-
-        desiredBalanceShardsAllocator.allocate(createAllocationFrom(createClusterState(discoveryNode, index2)), listener);
-
-        secondInputSubmitted.countDown();
-
-        terminate(threadPool);
-        assertThat(listenersCalled.get(), equalTo(2));
-    }
-
-    public void testFailListenersOnNoLongerMasterException() {
-
-        var threadPool = new TestThreadPool(getTestName());
-
-        var listenersCalled = new AtomicInteger(0);
+        };
 
-        RerouteService rerouteService = (r, p, l) -> l.onFailure(new NotMasterException("no longer master"));
+        var desiredBalanceShardsAllocator = DesiredBalanceShardsAllocator.create(
+            allocator,
+            threadPool,
+            clusterService,
+            rerouteServiceSupplier::get
+        );
+        var allocationService = new AllocationService(new AllocationDeciders(List.of()), new GatewayAllocator() {
+            @Override
+            public void beforeAllocation(RoutingAllocation allocation) {}
 
-        var desiredBalanceShardsAllocator = new DesiredBalanceShardsAllocator(new ShardsAllocator() {
             @Override
-            public void allocate(RoutingAllocation allocation) {
-                final var dataNodeId = allocation.nodes().getDataNodes().values().iterator().next().getId();
-                final var unassignedIterator = allocation.routingNodes().unassigned().iterator();
-                while (unassignedIterator.hasNext()) {
-                    unassignedIterator.next();
-                    unassignedIterator.initialize(dataNodeId, null, 0L, allocation.changes());
-                }
+            public void allocateUnassigned(
+                ShardRouting shardRouting,
+                RoutingAllocation allocation,
+                UnassignedAllocationHandler unassignedAllocationHandler
+            ) {
+                unassignedAllocationHandler.initialize(allocation.nodes().getLocalNodeId(), null, 0L, allocation.changes());
             }
 
             @Override
-            public ShardAllocationDecision decideShardAllocation(ShardRouting shard, RoutingAllocation allocation) {
-                throw new AssertionError("only used for allocation explain");
-            }
-        }, threadPool, () -> rerouteService);
+            public void afterPrimariesBeforeReplicas(RoutingAllocation allocation) {}
+        }, desiredBalanceShardsAllocator, () -> ClusterInfo.EMPTY, () -> SnapshotShardSizeInfo.EMPTY);
 
-        var discoveryNode = createDiscoveryNode();
-        var index = createIndex(UUIDs.randomBase64UUID());
+        rerouteServiceSupplier.set((r, p, l) -> {
+            clusterService.submitUnbatchedStateUpdateTask("test-desired-balance-reroute", new ClusterStateUpdateTask() {
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+                    logger.info("Performing reroute [{}]", r);
+                    reroutedIndices.addAll(createdIndices);
```

Comment:
Is `createdIndices` equal to the the list of indices in `currentState`? If so can we assert that?